### PR TITLE
14 - GitHub integration

### PR DESCRIPTION
**Pull request by jdesiloniz**
Fixes: #14 

This PR introduces the `compareBenchmarksCI` command to the plugin to allow integration with GitHub. When executed, enchmarks comparison are outputted both to the console and also to a comment in the specified PR. If any of the comparisons fail beyond the expected threshold, the mergeability status of the PR will be marked as `error` until a correct benchmark is submitted again.

### CODE REVIEW

File: modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubModel.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      checkOptionalParameter(accessToken, "gitHubToken"),
+      checkOptionalParameter(repositoryOwner, "repositoryOwner"),
+      checkOptionalParameter(repositoryName, "repositoryName"),
+      checkOptionalParameter(pullRequestNumber, "pullRequestNumber")
+    ).mapN {
+      case (token, repoOwner, repoName, pull) =>
+        GitHubParameters(token, repoOwner, repoName, pull, targetUrl)
+    }.toEither
+
+  private[this] def checkOptionalParameter[A](param: Option[A], name: String): ValidationResult[A] =
```
Comment (fedefernandez) [inline modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubModel.scala]: It's far from important, but should we rename the function to something like `checkMandatoryParameter` or something like that? We're returning an `Invalid` when is `None` but the name suggests something different.

File: modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubService.scala [hunk 2]
```diff
@@ -95,6 +107,40 @@ object GithubService {
...
+        targetUrl: Option[String],
+        description: String,
+        context: String
+    ): F[GHResponse[Status]] = {
+      val gh = Github(Some(accessToken))
+
+      (for {
+        pr <- EitherT(gh.pullRequests.get(owner, repository, pullRequestNumber).exec())
+        head <- EitherT.fromOption[F](
+          pr.result.head,
```
Comment (fedefernandez) [inline modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubService.scala]: This `head` looks very confusing, what is the type of `pr.result`?

File: modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala [hunk 3]
```diff
@@ -96,6 +136,40 @@ object SbtHoodPlugin extends AutoPlugin with SbtHoodDefaultSettings with SbtHood
...
   }
 
+  private def submitResultsToGitHub[F[_]](
+      benchmarkResult: List[BenchmarkComparisonResult],
+      previousPath: File,
+      currentPath: File,
+      params: GitHubParameters)(
+      implicit L: Logger[F],
+      S: Sync[F]): EitherT[F, GHException, Unit] = {
+    val gh = GithubService.build[F]
```
Comment (fedefernandez) [inline modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala]: Maybe we could move this to a `lazy` val in the `SbtHoodPlugin` object initialized over `IO`

File: modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubService.scala (hunk 2 above)
Comment (jdesiloniz) [reply modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubService.scala]: `head` refers to the head commit from the pullRequest (`result` is a GitHub4s `PullRequest` object). Statuses are associated to commits in a PR so I guessed the head commit would be the best option for this 😄

File: modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala (hunk 3 above)
Comment (jdesiloniz) [reply modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala]: Thanks for the suggestion! Done 😄

File: modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubModel.scala (hunk 1 above)
Comment (jdesiloniz) [reply modules/core/src/main/scala/com.fortysevendeg.hood/github/GithubModel.scala]: You're right, changed!

**Comment by jdesiloniz** [comment]
@fedefernandez Addressed your comments, could you please take a look again? Thanks! :D

**Pull request merged**